### ISSUE TITLE
`SCYLLA_TIMEUUID_LIST_INDEX(?)` confusingly accepts a `uuid` and not `timeuuid` as a parameter

### ISSUE THREAD

**Issue by piodul** [scylladb/scylladb#19918]
ScyllaDB has an extension to the CQL syntax which allows setting cells in list values directly by their timeuuid (described [here](https://github.com/scylladb/scylladb/blob/master/docs/dev/cql-extensions-internal.md#scylla_timeuuid_list_index-function)).

Consider this example:

```python
from cassandra.cluster import Cluster

s = Cluster().connect()

s.execute("CREATE KEYSPACE IF NOT EXISTS ks WITH replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 1} AND tablets = {'enabled': false}")
s.execute("CREATE TABLE IF NOT EXISTS ks.t (pk int, ck int, v1 list<int>, v2 timeuuid, PRIMARY KEY (pk, ck)) WITH cdc = {'enabled': true}")

p1 = s.prepare("UPDATE ks.t SET v1[SCYLLA_TIMEUUID_LIST_INDEX(?)] = 0, v2 = ? WHERE pk = 0 AND ck = 0")

print(p1.column_metadata[0].type)
print(p1.column_metadata[1].type)

p2 = s.prepare("SELECT * FROM ks.t_scylla_cdc_log WHERE v1 = ? ALLOW FILTERING")

print(p2.column_metadata[0].type)
```

Output:

```
<class 'cassandra.cqltypes.UUIDType'>
<class 'cassandra.cqltypes.TimeUUIDType'>
<class 'cassandra.io.libevreactor.MapType(TimeUUIDType, Int32Type)'>
```

Confusingly, `SCYLLA_TIMEUUID_LIST_INDEX` accepts a `uuid` parameter (first line in the output). This is a different CQL type than a `timeuuid` (as demonstrated by the second bind marker of the first query, and the second line in the output).

This is inconsistent and slightly inconvenient for CDC replicator applications which take data from CDC log and put it into another table. In CDC log, lists are represented as a map from timeuuids to the list element type (as evidenced by the third line in the output). Taking such a key from query results and binding it to a statement like `UPDATE ... SET v[SCYLLA_TIMEUUID_LIST_INDEX(?)] = ...` may require doing the `timeuuid` -> `uuid` conversion in runtime if the driver uses separate types to represent those two CQL types and does not accept one type in place of the other - specifically, this is the case for the rust driver.

Found in https://github.com/scylladb/scylla-cdc-rust/pull/125#discussion_r1690156257

**Comment by avikivity** [comment]
It will be a pain to fix due to compatibility (alternatively: difficult to come up with a new name for the function that accepts timeuuid).

Maybe we can make it polymorphic, preferring uuid, but able to accepted timeuuid via a hint: SCYLLA_TIMEUUID_LIST_INDEX((timeuuid)?).

**Comment by ptrsmrn** [comment]
@piodul assigning the `P4` label means we'll never do it :'-) But it's still good to have the issue opened to document and keep track of bugs.

**Comment by piodul** [comment]
> @piodul assigning the `P4` label means we'll never do it :'-) But it's still good to have the issue opened to document and keep track of bugs.

Yeah, I'm fully aware of this. The impact is low (very niche feature + there is a simple workaround) and as Avi wrote it might not be easy to fix.
